sisl/sparse.py

```diff
@@ -1323,24 +1358,35 @@ def tocsr(self, dim=0, **kwargs):
     def fromsp(cls, *sps, **kwargs):
         sps = list(map(lambda x: x.tocsr(), sps))
         shape = sps[0].shape
-        dtype = kwargs.get("dtype", np.result_type(*tuple(sp.dtype for sp in sps)))
-        # create a single sparse matrix to get the final size
-        # this will also error out if the shapes are not coherent
-        full_sp = reduce(lambda x, y: x + y, sps, 0.)
-        nnzpr = np.diff(full_sp.indptr).max() + 3
-        out = cls(shape + (len(sps), ), nnzpr=nnzpr, dtype=dtype)
-        del full_sp
-
-        # Now we need to add things to the sparsity pattern
-        for iD, sp in enumerate(sps):
-            ptr = sp.indptr
-            if sp.shape != shape:
-                raise ValueError(f"{cls.__name__}.fromsp found non compatible shapes")
-
-            # Loop stuff
-            for r in range(shape[0]):
-                idx = out._extend(r, sp.indices[ptr[r]:ptr[r+1]])
-                out._D[idx, iD] += sp.data[ptr[r]:ptr[r+1]]
+
+        if len(sps) == 1:
+            m = sps[0]
+            m.sort_indices()
+            out = cls(shape + (1,), nnzpr=1, nnz=1, dtype=m.dtype)
+            out.col = m.indices
+            out.ptr = m.indptr
+            out.ncol = np.diff(out.ptr)
+            out._nnz = len(out.col)
+            out._D = m.data.reshape(-1, 1)
+            if kwargs.get("copy", True):
+                out.col = out.col.copy()
+                out.ptr = out.ptr.copy()
+                out._D = out._D.copy()
+            return out
+        
+        # Pre-allocate by finding sparsity pattern union
+        kw1 = kwargs.copy()
+        kw1["copy"] = False
+        sps = [cls.fromsp(m, **kw1) for m in sps]
```

Comment:
I don't think you need this? First converting to `csr_matrix` then to `SparseCSR`?

You should directly interact with the `csr_matrix` I think this should be even faster?